### PR TITLE
feat(22.04): add crun and uidmap slices

### LINKED ISSUE #221: Create crun and uidmap slice
**ubuntu:22.04**

```
root@abf2f5b09a8f:/# dpkg -I /var/cache/apt/archives/crun_0.17+dfsg-1.1_riscv64.deb 
 new Debian package, version 2.0.
 size 740384 bytes: control archive=767 bytes.
     657 bytes,    13 lines      control              
     382 bytes,     6 lines      md5sums              
 Package: crun
 Version: 0.17+dfsg-1.1
 Architecture: riscv64
 Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
 Original-Maintainer: Dmitry Smirnov <onlyjob@debian.org>
 Installed-Size: 5292
 Depends: libc6 (>= 2.34), libcap2 (>= 1:2.10), libseccomp2 (>= 2.2.1), libsystemd0 (>= 237), libyajl2 (>= 2.0.4)
 Built-Using: golang-github-opencontainers-image-spec (= 1.0.2-2), golang-github-opencontainers-specs (= 1.0.2.66.g20a2d97-1)
 Section: admin
 Priority: optional
 Homepage: https://github.com/containers/crun
 Description: lightweight OCI runtime for running containers
  Fast and low-memory footprint OCI Container Runtime fully written in C.
```

If I got things right, we'll need these slices definitions:

- [x] crun
- [x] libcap2
- [x] liblz4
- [x] libseccomp2
- [x] libsystemd0
- [x] libyajl2
- [x] libaudit-common
- [x] libaudit1
- [x] libcap-ng0
- [x] uidmap

I pretend to work on those.

### PR DESCRIPTION
**Pull request by endersonmaia**
# Proposed changes

Add crun and uidmap slices to 22.04.

Close #221 

### Forward porting

- https://github.com/canonical/chisel-releases/pull/223
- https://github.com/canonical/chisel-releases/pull/224
- https://github.com/canonical/chisel-releases/pull/225

## Testing

Will 

```bash
# on a local copy in the current branch run:
mkdir rootfs
chisel cut \
    --release ./ \
    --root rootfs/ \
    --arch=amd64 \
    base-files_base \
    base-files_release-info \
    crun_bins
cat <<EOF | docker build -t crun-container . -f -
FROM scratch
COPY rootfs/ /
EOF
docker run --rm crun-container crun --version
rm -rf rootfs/
```

Expected:

```
crun version 0.17
commit: 0e9229ae34caaebcb86f1fde18de3acaf18c6d9a
spec: 1.0.0
+SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
```

## Checklist

* [x] I have read the [contributing guidelines](
https://github.com/canonical/chisel-releases/blob/main/CONTRIBUTING.md)
* [x] I have already submitted the [CLA form](
https://ubuntu.com/legal/contributors/agreement)

### CODE REVIEW

**Review by zhijie-yang — CHANGES_REQUESTED** [review]
Looks good. There are some issues regarding the linting (see workflow runs) to be fixed, and it is suggested to add the copyright files explicitly in the slice definitions.

**Review by linostar — COMMENTED** [review]
slice names to sorted, and copyright need to be added:
- crum: `/usr/share/doc/crun/copyright:`
- libaudit-common: `/usr/share/doc/libaudit-common/copyright:`
- libaudit1: `/usr/share/doc/libaudit1/copyright:`
- libcap-ng0: `/usr/share/doc/libcap-ng0/copyright:`
- libcap2: `/usr/share/doc/libcap2/copyright:`
- liblz4-1: `/usr/share/doc/liblz4-1/copyright:`
- libseccomp2: `/usr/share/doc/libseccomp2/copyright:`
- libsystemd0: `/usr/share/doc/libsystemd0/copyright:`
- libyajl2: `/usr/share/doc/libyajl2/copyright:`
- uidmap: `/usr/share/doc/uidmap/copyright:`

**Review by linostar — APPROVED** [review]
LGTM, thanks!

File: slices/crun.yaml
Comment (zhijie-yang) [inline slices/crun.yaml:1]: Message from `crun` copyright file (`/usr/share/doc/crun/copyright`): Please don't forget to add me here!

File: slices/uidmap.yaml
Comment (zhijie-yang) [inline slices/uidmap.yaml:1]: `uidmap`: So it my copyright file! It's here -> `/usr/share/doc/uidmap/copyright`!

**Review by zhijie-yang — CHANGES_REQUESTED** [review]

File: slices/crun.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  bins:
+    essential:
+      - libc6_libs
+      - libcap2_libs
+      - libseccomp2_libs
+      - libsystemd0_libs
+      - libyajl2_libs
+    contents:
+      /usr/bin/crun:
+      /usr/lib/*-linux-*/libcrun.a:
```
Comment (rebornplusplus) [inline slices/crun.yaml:16]: This file should go in a separate `libs` slice. The slice then needs to be listed as an essential of `bins`.

File: slices/libgcrypt20.yaml [hunk 2]
```diff
@@ -11,7 +11,6 @@ slices:
       - libgpg-error0_libs
     contents:
       /usr/lib/*-linux-*/libgcrypt.so.20*:
-
```
Comment (rebornplusplus) [inline slices/libgcrypt20.yaml:14]: We don't need to change existing files blank spaces. Same for below.

**Review by rebornplusplus — COMMENTED** [review]
The slices look good, thanks! Only left a few comments below. 

With the recent changes around testing in the upstream branch, I think it would be nice to have a spread test for `crun` and `uidmap` as well.

**Review by rebornplusplus — APPROVED** [review]
Updated the trivial changes myself. Looks good to me, thank you!

Tried to add a smoke test by running `crun --version` in a `chroot` env, but `crun` runs into an error. Probably this is related: https://github.com/containers/crun/issues/1434. Reverted the commit I added for smoke test. Please feel free to add a smoke test of your own!

**Comment by cjdcordeiro** [comment]
@rebornplusplus this is actually a very similar issue to #274 -> `crun` needs procfs, which doesn't exist in the chroot env and we can't create it cause the test runs inside an unprivileged container.

If we can't find a workaround until tomorrow, let's just add a `TODO` comment in the SDF and move on.

**Comment by cjdcordeiro** [comment]
@endersonmaia how urgent is this? I'm trying to find a different testing strategy to support such privileged tests like yours, but it could take a few days. If you need this now, I'm happy to merge it with the simplified tests and improve those later on.

**Comment by endersonmaia** [comment]
> @endersonmaia how urgent is this? I'm trying to find a different testing strategy to support such privileged tests like yours, but it could take a few days. If you need this now, I'm happy to merge it with the simplified tests and improve those later on.

I'm using my fork's branch to have this in my build pipeline, so it's not urgent.

**Comment by endersonmaia** [comment]
> ... I'm trying to find a different testing strategy to support such privileged tests like yours ...

Have you considered running qemu or even firecracker?

I use the [crane](https://github.com/google/go-containerregistry/blob/main/cmd/crane/doc/crane_export.md) tool to generate a rootfs and together with a kernel image you could start a firecracker microVM using something like [firectl](https://github.com/firecracker-microvm/firectl)

**Comment by cjdcordeiro** [comment]
Thanks @endersonmaia and thanks for the suggestions. I already have #317 up to introduce LXD as a backend for the tests execution. I've tested it with your PR and seemed to work just fine. So once it's merged, we can rebase your PR.

If you have some tests, please add them and I'll re-trigger the CI after the rebase

**Comment by cjdcordeiro** [comment]
> Updated the trivial changes myself. Looks good to me, thank you!
> 
> Tried to add a smoke test by running `crun --version` in a `chroot` env, but `crun` runs into an error. Probably this is related: [containers/crun#1434](https://github.com/containers/crun/issues/1434). Reverted the commit I added for smoke test. Please feel free to add a smoke test of your own!

@rebornplusplus (or @endersonmaia ) could you please add such test here (and counterpart releases)? :pray:

**Comment by cjdcordeiro** [comment]
Alright, I've added the `crun` test. Looking good. Merging.

**Review by cjdcordeiro — APPROVED** [review]

**Pull request merged**